lib/mongo/collection/view/change_stream.rb

```diff
@@ -270,17 +270,22 @@ def create_cursor!
           @start_at_operation_time_supported = nil
 
           session = client.send(:get_session, @options)
-          server = server_selector.select_server(cluster)
-          result = send_initial_query(server, session)
-          if doc = result.replies.first && result.replies.first.documents.first
-            @start_at_operation_time = doc['operationTime']
-          else
-            # The above may set @start_at_operation_time to nil
-            # if it was not in the document for some reason,
-            # for consistency set it to nil here as well
-            @start_at_operation_time = nil
+          start_at_operation_time = nil
+          @cursor = read_with_retry_cursor(session, server_selector, view) do |server|
+            result = send_initial_query(server, session)
+            if doc = result.replies.first && result.replies.first.documents.first
+              start_at_operation_time = doc['operationTime']
+            else
+              # The above may set @start_at_operation_time to nil
+              # if it was not in the document for some reason,
+              # for consistency set it to nil here as well.
+              # NB: since this block may be executed more than once, each
```

Comment:
Will the block be executed more than once if the first branch of the `if/else` is taken? My assumption is that if the driver successfully receives a reply with an operation time, then the read was successful and it won't be retried. If that's the case, why do you we need to set the operation time to `nil` here if it's set to `nil` above the call to `read_with_retry_cursor`?

Reply:
It is possible that the cursor creation might raise an error which will cause the read to be retried, for example if a future version of the driver decides to communicate with the server during cursor creation. Since an assignment is inexpensive, being defensive and clearing `start_at_operation_time` makes this method not reliant on behavior elsewhere in the driver to function correctly.

Reply:
I'm not sure how raising an error during cursor creation could cause this block to have `start_at_operation_time` populated with a non-nil value (given that the error would be raised during the `result.replies.first && result.replies.first.documents.first` check and therefore wouldn't even get to the part where the operation time is set, and I don't think "relying on behavior elsewhere in a the driver" is a very meaningful description of "relying on the condition of the `if` branch that this `else` block is attached to, but I doubt we'll come to any agreement about this, and it's not super important.